chairman/Chairman.hs

```diff
@@ -0,0 +1,434 @@
+
+
+--
+-- Client Application
+--
+
+localInitiatorNetworkApplication
+  :: forall blk m peer.
+     ( RunNode blk
+     , Condense blk
+     , MonadAsync m
+     , MonadSTM   m
+     , MonadST    m
+     , MonadThrow m
+     , MonadTimer m
+     )
+  => CoreNodeId
+  -> ChainsVar m blk
+  -> Int
+  -> Maybe BlockNo
+  -> Tracer m (ChairmanTrace blk)
+  -> Tracer m (TraceSendRecv (ChainSync blk (Point blk)) peer DeserialiseFailure)
+  -- ^ tracer which logs all chain-sync messages send and received by the client
+  -- (see 'Ouroboros.Network.Protocol.ChainSync.Type' in 'ouroboros-network'
+  -- package)
+  -> Tracer m (TraceSendRecv (LocalTxSubmission (GenTx blk) String) peer DeserialiseFailure)
+  -- ^ tracer which logs all local tx submission protocol messages send and
+  -- received by the client (see 'Ouroboros.Network.Protocol.LocalTxSubmission.Type'
+  -- in 'ouroboros-network' package).
+  -> NodeConfig (BlockProtocol blk)
+  -> Versions NodeToClientVersion DictVersion
+              (OuroborosApplication InitiatorApp peer NodeToClientProtocols
+                                    m ByteString () Void)
+localInitiatorNetworkApplication coreNodeId chainsVar securityParam maxBlockNo chairmanTracer chainSyncTracer localTxSubmissionTracer pInfoConfig =
+    simpleSingletonVersions
+      NodeToClientV_1
+      (NodeToClientVersionData { networkMagic = 0 })
+      (DictVersion nodeToClientCodecCBORTerm)
+
+  $ OuroborosInitiatorApplication $ \peer ptcl -> case ptcl of
+      LocalTxSubmissionPtcl -> \channel -> do
+        txv <- newEmptyTMVarM @_ @(GenTx blk)
+        runPeer
+          localTxSubmissionTracer
+          localTxSubmissionCodec
+          peer
+          channel
+          (localTxSubmissionClientPeer
+              (txSubmissionClient @(GenTx blk) txv))
+
+      ChainSyncWithBlocksPtcl -> \channel ->
+        runPeer
+          chainSyncTracer
+          (localChainSyncCodec pInfoConfig)
+          peer
+          channel
+          (chainSyncClientPeer $ chainSyncClient chairmanTracer coreNodeId chainsVar securityParam maxBlockNo)
+
+
+--
+-- Submission client
+--
+
+-- | A 'LocalTxSubmissionClient' that submits transactions reading them from
+-- a 'TMVar'.  A real implementation should use a better synchronisation
+-- primitive.  This demo creates and empty 'TMVar' in
+-- 'muxLocalInitiatorNetworkApplication' above and never fills it with a tx.
+--
+txSubmissionClient
+  :: forall tx reject m.
+     ( Monad    m
+     , MonadSTM m
+     )
+  => TMVar m tx
+  -> m (LocalTxSubmissionClient tx reject m ())
+txSubmissionClient txv = do
+    tx <- atomically $ readTMVar txv
+    pure $ SendMsgSubmitTx tx $ \mbreject -> do
+      case mbreject of
+        Nothing -> return ()
+        Just _r -> return ()
+      txSubmissionClient txv
+
+--
+-- Chain-Sync client
+--
+
+-- | 'ChainSyncClient' which build chain fragment; on every roll forward it will
+-- check if there is consensus on immutable chain.
+--
+chainSyncClient
```

Comment:
Can you move this up a bit, closer to the implementation of the helper functions it uses (`addBlock`, etc.)? Because I constantly had to scroll back and forth over unrelated code :slightly_smiling_face: